### PR TITLE
Check Fluid Amounts and Other Updates.

### PR DESCRIPTION
**Pull request by AmirAbrams**
### What is the purpose of this pull request (PR)?
- Fix block reward debug output logging
- Update verbiage in debug log and add missing ENABLE_WALLET comment
- Remove existing dDNS code
- Check if exceeds maximum fluid amount and if negative amount.  Enforce with consensus checking.
dynamic-fluid-tx-check_20171207

- Check if fluid transaction is already in the memory pool
dynamic-fluid-mem-pool-check2_20171208

### CODE REVIEW

File: src/fluid.cpp [hunk 1]
```diff
@@ -74,6 +74,56 @@ bool Fluid::IsGivenKeyMaster(CDynamicAddress inputKey) {
     return false;
 }
 
+/** Checks fluid transactoin operation script amount for invalid values. */
+bool Fluid::CheckFluidOperationScript(const CScript& fluidScriptPubKey, std::string& errorMessage) {
+    std::string strFluidOpScript = ScriptToAsmStr(fluidScriptPubKey);
+    std::string verificationWithoutOpCode = GetRidOfScriptStatement(strFluidOpScript);
+    std::string strOperationCode = GetRidOfScriptStatement(strFluidOpScript, 0);
+    if (!ExtractCheckTimestamp(strFluidOpScript, GetTime())) {
+        errorMessage = "CheckFluidOperationScript fluid timestamp is too old.";
```
Comment (ghost) [inline src/fluid.cpp]: Range can be in past and in future, should be "out of range" error

File: src/fluid.cpp [hunk 2]
```diff
@@ -74,6 +74,56 @@ bool Fluid::IsGivenKeyMaster(CDynamicAddress inputKey) {
     return false;
 }
 
+/** Checks fluid transactoin operation script amount for invalid values. */
+bool Fluid::CheckFluidOperationScript(const CScript& fluidScriptPubKey, std::string& errorMessage) {
```
Comment (ghost) [inline src/fluid.cpp]: Similar restrictions must be put onto RPC Interface to prevent reaching this stage in the first place

File: src/fluid.h [hunk 3]
```diff
@@ -29,7 +29,7 @@ class CFluidParameters {
     static const int FLUID_ACTIVATE_HEIGHT = 10;
     static const int64_t MAX_FLUID_TIME_DISTORT = 60 * 60; // Maximum time distort = 1 hour.
     static const CAmount FLUID_TRANSACTION_COST = 100000 * COIN; // Cost to send a fluid transaction
-    static const CAmount FLUID_MAX_REWARD_FOR_DYNODE_ = 1000 * COIN; // Max dynode block reward using fluid OP_REWARD_DYNODE
+    static const CAmount FLUID_MAX_REWARD_FOR_DYNODE = 1000 * COIN; // Max dynode block reward using fluid OP_REWARD_DYNODE
     static const CAmount FLUID_MAX_REWARD_FOR_MINING = 1000 * COIN; // Max mining block reward using fluid OP_REWARD_MINING
     static const CAmount FLUID_MAX_FOR_MINT = 1000000000 * COIN; // Max minting amount per fluid transaction
```
Comment (ghost) [inline src/fluid.h:35]: Should we just set this to std::max (int64_t)?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3243,6 +3244,13 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
                 {
                     CTxOut txout(recipient.nAmount, recipient.scriptPubKey);
                     
+                    if (IsTransactionFluid(recipient.scriptPubKey)) {
+                        // Check the validity of the fluid transaction's public script.
+                        if (!fluid.CheckFluidOperationScript(recipient.scriptPubKey, strFailReason)) {
```
Comment (ghost) [inline src/wallet/wallet.cpp]: Same check to be done in CheckTransaction in validation.cpp

**Review by ghost — COMMENTED** [review]
Overall seems well and good, some recommendations commented, good work!

File: src/fluid.h (hunk 3 above)
Comment (AmirAbrams) [reply src/fluid.h:35]: We tested a 9223372036854775806 DYN.  When it tried to add to that number, it crashes so we implemented max limits.

File: src/fluid.cpp (hunk 1 above)
Comment (AmirAbrams) [reply src/fluid.cpp]: Yeah, you are correct.  I think I should change GetTime() to the transaction time.  It is causing this error:
`ERROR: CheckBlock(): CheckFluidOperationScript fluid timestamp is too old....`

File: src/fluid.cpp (hunk 2 above)
Comment (AmirAbrams) [reply src/fluid.cpp]: Yes, I agree it is best to stop the transaction before adding it to the wallet and broadcasting/committing it.  

I added CheckFluidOperationScript to CWallet::CreateTransaction to stop the RPC command before committing the transaction.  
dynamic-fluid-tx-check_20171207

File: src/fluid.cpp (hunk 1 above)
Comment (AmirAbrams) [reply src/fluid.cpp]: I changed it to a timestamp and it fixed the issue.  thanks again.

File: src/fluid.h (hunk 3 above)
Comment (ghost) [reply src/fluid.h:35]: Makes sense... Thanks!

**Pull request merged**